### PR TITLE
add lastAssignedFieldId to table metadata

### PR DESCRIPTION
**Pull request by jun-he**
Follow up https://github.com/apache/iceberg/pull/2083.
Add `lastAssignedFieldId` to table metadata. 
To keep it simple, `lastAssignedFieldId` across all partition specs will be computed at initialization without persisting to the JSON file.

### CODE REVIEW

**Review by yyanyy — APPROVED** [review]
Thanks for making the change! LGTM.

One question: do you intend to add `lastAssignedFieldId` to the JSON file in another PR, or you don't think it's necessary (so that this change is mostly to make `lastAssignedFieldId` available in table metadata class)? I thought we want to persist it in JSON file, and if that's the case, I'll echo my question from [this PR](https://github.com/apache/iceberg/pull/2083#issuecomment-759833051): 
> Is this just to ensure implementation correctness per spec? As it seems that from the current implementation getting this from PartitionSpec::lastAssignedFieldId should be good enough, as there's no table operation to actually remove a partition field from specs?

**Comment by jun-he** [comment]
@yyanyy Thanks for the review. 
I thought about adding it to metadata JSON file. It seems unnecessary and also adds additional complexity for backward compatibility and maintains its correctness. As the info is already included in specs, we can simply recompute it during initialization.

It seems table metadata is a better place to hold this info than partition spec updater as it is an info for a table, which might be used by other code paths in the future.

**Comment by rdblue** [comment]
The last assigned partition field ID does need to be added to the spec and the table metadata JSON. It is possible to reuse IDs, but I don't think it is a good idea to do it. There is also no need to reuse them. I think that reuse could only cause problems.

**Comment by rdblue** [comment]
I just thought of an example where ID reuse could be bad. The Nessie catalog allows branching at the table metadata level by keeping different metadata file references. A Nessie branch could remove data for one spec and reuse the IDs, which would make compatibility with another branch using the old spec confusing.

**Comment by jun-he** [comment]
@rdblue Thanks for the comment. 
It makes sense to add the lastAssignedFieldId to metadata JSON when specs in metadata might be removed.
We already keep the lastAssignedFieldId in the spec JSON. I will add lastAssignedFieldId to table metadata JSON as well.

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java
Comment (yyanyy) [inline core/src/main/java/org/apache/iceberg/TableMetadataParser.java:296]: Do we want it to be a required field in v2 and fail validation if it's missing?

**Review by yyanyy — COMMENTED** [review]
LGTM, just one comment

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java
Comment (jun-he) [reply core/src/main/java/org/apache/iceberg/TableMetadataParser.java:296]: It makes sense as the V2 is not released yet. I will update it accordingly.

**Comment by jun-he** [comment]
Will update the doc after the change is merged.

**Review by yyanyy — APPROVED** [review]

**Comment by jun-he** [comment]
@rdblue can you please take another look? thanks.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:377]: I think this name is potentially confusing. It is the last assigned partition field ID, and "partition" is a key part of it. How about renaming it to `lastAssignedPartitionId` instead?

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadataParser.java:297]: Should this use a precondition to check that the last assigned partition ID is present? That would result in a better error message if it is missing, like the other cases: `last-assigned-partition-id must exist in format v2`

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadataParser.java:95]: I think this should be `last-assigned-partition-id`.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:719]: I think that we need to fix the implementation of `freshSpec` because it doesn't accept a `lastAssignedFieldId` or reuse existing partition field IDs. In fact, I think what would currently happen is that the spec would be created independent of the rest of the table metadata and could reuse existing IDs. That's not good, but we also don't want to reassign all of the IDs (though that would be less bad).

We need to create a new spec similar to how we use `assignFreshIds` for the schema, which will reuse IDs for existing fields and only assign new IDs for new fields, starting at the last assigned ID.

**Comment by rdblue** [comment]
Thank you, @jun-he! I think there are few main improvements needed:
1. We should use `lastAssignedPartitionId` instead of `lastAssignedFieldId` in all cases to be more specific
2. For a table replace, we need to reassign the partition field IDs and reuse any existing IDs for equivalent fields
3. We need to update the spec doc

We can do all 3 in this PR, or just the first one if you prefer to separate them. Just let me know.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java
Comment (jun-he) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java]: Updated the field name in `TableMetadata`. 
In `PartitionSpec`, there is also a `lastAssignedFieldId` field. It seems to be fine as it is a field inside partition spec, which already indicates it is about partition.

**Comment by jun-he** [comment]
@rdblue  thanks for the comments. Updated the PR with the renamed field. I will send separate PRs for item 2 and 3. Thanks.

File: core/src/test/resources/TableMetadataV2MissingLastAssignedPartitionId.json
Comment (rdblue) [inline core/src/test/resources/TableMetadataV2MissingLastAssignedPartitionId.json:7]: I just realized that this is `last-column-id` and not `last-assigned-column-id`. I think we should make the metadata property `last-partition-id` instead of `last-assigned-partition-id` to match. I'll merge this and we can fix it in a follow-up.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @jun-he!

**Pull request merged**
